### ISSUE TITLE
Geo search

### ISSUE THREAD

**Issue by rpostulart** [aws-amplify/amplify-js#2475]
At this moment there is no possibility to search via graphql / dynamodb certain records based on the longitude and latitude (geo) while there is a kind of library within AWS for this: 
https://github.com/amazon-archives/dynamodb-geo

Can we make these functionalities also available via Amplify:
**Box Queries**: Return all of the items that fall within a pair of geo points that define a rectangle as projected onto a sphere.
**Radius Queries:** Return all of the items that are within a given radius of a geo point.
**Basic CRUD Operations:** Create, retrieve, update, and delete geospatial data items.
**Easy Integration:** Adds functionality to the AWS SDK for Java in your server application.
**Customizable:** Access to raw request and result objects from the AWS SDK for Java.

**Comment by doom777** [comment]
Elastic search is a workaround. We'd like native geo features, like Firebase is about to offer. Amplify is meant for mobile apps, and  "get all things in this geo box" or "get all things in a radius around this geo point"  is a very popular usage for mobiles apps. Please consider making this a high priority

**Comment by undefobj** [comment]
DynamoDB does not support geo queries, which is why that repo is in archive now. AppSync supports both DynamoDB as well as Elasticsearch which you can leverage in a single API and stream the data from DynamoDB to Elasticsearch indexes. That's why `@searchable` was added: https://aws-amplify.github.io/docs/cli/graphql#searchable

I'm not sure what the feature request is here since Amplify leverages the backend AWS services. Could you see if `@searchable` in the transformer meets your use case?

**Comment by temideoye** [comment]
@undefobj, the `@searchable` directive provides an interface for String, Int, Float, and Boolean searches. If I understand it correctly, I think the feature request here is for geo-related searches which is probably two requests in one when broken down:

**1.** Support geo-types, i.e a new GraphQLObjectType, like AWSGeo?
**2** Enable geo-related queries via `@searchable`, which is a bit more involved than your typical `eq`, `ne`, `lt`, `gt` e.t.c

**Comment by undefobj** [comment]
For new GraphQL scalar types that's best logged as a feature request to the AppSync service. I would recommend logging that here: https://forums.aws.amazon.com/forum.jspa?forumID=280&start=0

If there are enhancements needed to the GraphQL transformer it's best logged as a feature request in the Amplify CLI repo (as this repo is for the library): https://github.com/aws-amplify/amplify-cli

@rpostulart does this match with your request?

**Comment by doom777** [comment]
Why do you need ElasticSearch anyway for simple queries such as 'lt' 'ne' etc? DynamoDB supports these by itself. Is this just a away to force users to use and pay for ES even when not necessary?

Anyway, you can now do a Float lat and Float lon and do box queries by havin lat and lat lt and lon gt and lon lt, but come on... Also it fails at poles or if you need large boxes or.... Anyway you shouldn't have to be a spherical geometry expert to do mobiles apps, that's what libs are for (although basic knowledge is still a must IMHO)

**Comment by doom777** [comment]
@undefobj 
True dynamodb doesn't support geo features natively, so that archived lib simulated geo features by using geohashes and in-memory filtering, similar to what firebase geo libs do. This would be an acceptable solution fornus IMO, or else intergrate '@searchable' with ES geo queries and types as guys previously said.

**Comment by rpostulart** [comment]
@undefobj I will check asap and let it know here.

**Comment by doom777** [comment]
There is this [tutorial](https://serverless.com/blog/build-geosearch-graphql-api-aws-appsync-elasticsearch/) for geo queries in AppSync. Also, this AppSync [tutorial](https://docs.aws.amazon.com/appsync/latest/devguide/tutorial-combining-graphql-resolvers.html) mentions geospatial data/. So it is probably somehow possible, just very convoluted.

**Comment by undefobj** [comment]
@doom777 you can use `@model` to generate the simple queries that you state above, as outlined here: https://aws-amplify.github.io/docs/cli/graphql#generates

However these will just use DynamoDB itself and at scale these filtering techniques may be less performant based on your data access pattern which is why Elasticsearch exists. The `@searchable` will automatically stream data from DynamoDB to Elasticsearch like in the tutorial that you linked above.

**Comment by rpostulart** [comment]
I checked and agree with @doom777 . ElasticSearch does not suits my use case. I believe that ElasticSearch is too heavily for this use case. Because I expect it will be a highly used mobile app. I don't want to spin up an instance for each request, I prefer a serverless solutions for example with Lambda.

**Comment by doom777** [comment]
I may be wrong, but I don't think an ES instance is spun up on every request; I think it stays there permanently, possibly in a cluster.

**Comment by rpostulart** [comment]
Sorry that's what I mean indeed!

**Comment by undefobj** [comment]
Right, so to be clear @rpostulart as in the link I posted above `@model` will include some filtering capability out of the box with DynamoDB. For instance any `string` types will have operations like less than, between, etc:

```
input ModelStringFilterInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
}
```

This is probably good for prototyping & small apps, which is why we added this capability based on customer feedback. However, if your app takes off and your table gets very large this becomes less scalable and you'll want to leverage a purpose-built search engine like Elasticsearch which is why `@searchable` exists. The Elasticsearch instance is not "serverless" though as it will always be up whereas DynamoDB is a per-request model. If you do believe your app will be highly used then you should go this route, however if not then you could just start with the base DynamoDB capabilities and later you could always add in Elasticsearch. It's not a one-way door. AWS does not offer a "serverless full text search" service at this time so outside of this not much else we can offer here.

**Comment by rpostulart** [comment]
Ok thanks, then I will focus on that.

**Comment by doom777** [comment]
At the very least AWS could offer great documentation on how to implement this common usw casr

**Comment by undefobj** [comment]
@doom777 what more documentation could we add here? I linked to the transformer documentation above, is more needed there? Or alternatively is something else specifically needed?

**Comment by doom777** [comment]
Look, I need to do a "select all entity in this geo box" query. Please show me an AWS tutorial how to do it. This is a common enough case.

**Comment by doom777** [comment]
An AppSync dev has confirmed that this feature is in the planning pipeline
https://forums.aws.amazon.com/thread.jspa?threadID=296334&tstart=0
